pkg/ingester/checkpoint.go

```diff
@@ -59,3 +76,420 @@ func fromWireChunks(conf *Config, wireChunks []Chunk) ([]*chunkDesc, error) {
 	}
 	return descs, nil
 }
+
+// nolint:interfacer
+func decodeCheckpointRecord(rec []byte, s *Series) error {
+	//TODO(owen-d): reduce allocs
+	// The proto unmarshaling code will retain references to the underlying []byte it's passed
+	// in order to reduce allocs. This is harmful to us because when reading from a WAL, the []byte
+	// is only guaranteed to be valid between calls to Next().
+	// Therefore, we copy it to avoid this problem.
+	cpy := make([]byte, len(rec))
+	copy(cpy, rec)
+
+	switch RecordType(cpy[0]) {
+	case CheckpointRecord:
+		return proto.Unmarshal(cpy[1:], s)
+	default:
+		return errors.Errorf("unexpected record type: %d", rec[0])
+	}
+}
+
+func encodeWithTypeHeader(m proto.Message, typ RecordType, b []byte) ([]byte, error) {
+	buf, err := proto.Marshal(m)
+	if err != nil {
+		return b, err
+	}
+
+	b = append(b[:0], byte(typ))
+	b = append(b, buf...)
+	return b, nil
+}
+
+type SeriesWithErr struct {
+	Err    error
+	Series *Series
+}
+
+type SeriesIter interface {
+	Num() int
+	Iter() <-chan *SeriesWithErr
+	Stop()
+}
+
+type ingesterSeriesIter struct {
+	ing *Ingester
+
+	done chan struct{}
+}
+
+func newIngesterSeriesIter(ing *Ingester) *ingesterSeriesIter {
```

Comment:
I think you should abstract away with a interface defined here the ingester. This way you clearly show what you need and reduce the coupling.

This has also the advantages of making testing easier which I think you don't have on the `Iter()` part.

It might requires ingester code refactoring to properly expose what you need so we can keep this for later.

Reply:
The interface is defined as 
```go
type SeriesIter interface {
	Num() int
	Iter() <-chan *SeriesWithErr
	Stop()
}
```
here: https://github.com/grafana/loki/blob/316618342cfccb64b8db22498e0228be339a4644/pkg/ingester/checkpoint.go#L114-L118

Do you mean something else?

Reply:
yes I was talking about the Ingester passed in as Parameter in the new here.